pages/blog/_posts/2022-01-10-the-spec-behind-the-specs.md

```diff
@@ -0,0 +1,1347 @@
+    ...,
+    stop,
+    half
+} -- imaginary version 3
+```
+
+For `SEQUENCE`, `SET` and `CHOICE` one could either do the same, or to
+keep the versions separated, add another extension mark with the new
+fields in between. One could also add version brackets to group the
+extensions and highlight the differences.
+
+```
+Ax ::= SEQUENCE {
+    a INTEGER (250..253),
+    b BOOLEAN,
+    c CHOICE {
+        d INTEGER, -- version 1
+        ...,
+        [[
+            e BOOLEAN,
+            f IA5String
+        ]],
+        ... -- version 2
+    },
+    ..., -- version 1
+    [[
+        g NumericString (SIZE(3)),
+        h BOOLEAN OPTIONAL
+    ]],
+    ..., -- version 2
+    i BMPString OPTIONAL,  -- version 3
+    j PrintableString OPTIONAL -- version 3
+}
+
+```
+
+Only the types `ENUMERATED`, `SEQUENCE`, `SET` and `CHOICE`, as well
+as subtype constraints, and object and value sets can be extended.
+
+## Automatic, Implicit, Explicit tags
+
+When an value is transmitted all ambiguities need to be removed. That
+is why every type needs to have an unique identifier, called a tag.
+
+The section [DEFAULT and OPTIONAL
+keywords](#default-and-optional-keywords) has a good example I will
+explain.
+
+
+```
+CollectedDigits ::= SEQUENCE {
+  minimumNbOfDigits    [0] INTEGER (1..16) DEFAULT 1,
+  maximumNbOfDigits    [1] INTEGER (1..16),
+  endOfReplyDigit      [2] OCTET STRING (SIZE (1..2)) OPTIONAL,
+  cancelDigit          [3] OCTET STRING (SIZE (1..2)) OPTIONAL,
+  startDigit           [4] OCTET STRING (SIZE (1..2)) OPTIONAL,
+  firstDigitTimeOut    [5] INTEGER (1..127) OPTIONAL,
+  interDigitTimeOut    [6] INTEGER (1..127) OPTIONAL,
+  errorTreatment       [7] ErrorTreatment DEFAULT stdErrorAndInfo,
+  interruptableAnnInd  [8] BOOLEAN DEFAULT TRUE,
+  voiceInformation     [9] BOOLEAN DEFAULT FALSE,
+  voiceBack            [10] BOOLEAN DEFAULT FALSE
+}
+```
+
+In this sequence we can see that many of the values are optional and
+some have defaults, only the value `maximumNbOfDigits` is mandatory.
+
+When the sending side transmits a value of `CollectedDigits` type, the
+receiving side will get a sequence of the values mentioned. With
+BER-encoding each defined value will have an identifier, a length (of
+the value transmitted) and the value. This is called a
+Tag-Length-Value or TLV for short.
+
+All basic types already has an universal tag as stated in the [types
+table](#types), but the composite types does not. (If not tagged, how
+would it see the difference between `endOfReplyDigit`, `cancelDigit`
+and `startDigit` in the example above?)
+
+When explicitly tagged (as above) the both the tag `[2]` and the
+universal tag `[4]` for `OCTET STRING` would be transmitted for
+`endOfReplyDigits`. This is all fine and well, because then the types
+could be potentially extended, the `OCTET STRING` could for instance
+be replaced with a `BIT STRING` and the receiver would still
+understand the message because the universal tag `[3]` would be sent
```

Comment:
Hm, I don't know if this is realistic. The receiver code was most likely created from the ASN.1 spec which specified an OCTET STRING.

Reply:
Removed